src/compose/MentionWarnings.js

```diff
@@ -0,0 +1,162 @@
+/* @flow strict-local */
+
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+
+import type { Auth, Stream, Dispatch, Narrow, UserOrBot, Subscription } from '../types';
+import { getActiveUsersById, getAuth } from '../selectors';
+import { isPrivateNarrow } from '../utils/narrow';
+import * as api from '../api';
+
+import AnimatedScaleComponent from '../animation/AnimatedScaleComponent';
+import MentionedUserNotSubscribed from '../message/MentionedUserNotSubscribed';
+
+type State = {|
+  unsubscribedMentions: Array<number>,
+|};
+
+type SelectorProps = {|
+  auth: Auth,
+  usersById: Map<number, UserOrBot>,
+|};
+
+type Props = $ReadOnly<{|
+  narrow: Narrow,
+  stream: Subscription | {| ...Stream, in_home_view: boolean |},
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class MentionWarnings extends PureComponent<Props, State> {
+  state = {
+    unsubscribedMentions: [],
+  };
+
+  /**
+   * Tries to parse a user object from an @-mention.
+   *
+   * @param completion The autocomplete option chosend by the user.
+      See JSDoc for AutoCompleteView for details.
+   */
+  getUserFromMention = (completion: string): UserOrBot | void => {
+    const { usersById } = this.props;
+
+    const unformattedMessage = completion.split('**')[1];
+    const [userFullName, userIdRaw] = unformattedMessage.split('|');
+
+    // We skip user groups, for which autocompletes are of the form
+    // `*<user_group_name>*`, and therefore, message.split('**')[1]
+    // is undefined.
+    if (unformattedMessage === undefined) {
+      return undefined;
+    }
+
+    if (userIdRaw !== undefined) {
+      const userId = Number.parseInt(userIdRaw, 10);
+      const user = usersById.get(userId);
+      if (user !== undefined) {
+        return user;
```

Comment:
Let's leave out the `if (user !== undefined)` and just unconditionally return `user` at this point.

Consider what that condition means:
* There was an explicit user ID in the @-mention.
* For whatever reason, we don't find the user that the @-mention specifically identified by user ID.
* We don't return yet, fall through, and end up looking to see if there's some *other* user who happens to have the same name, and return that one.

If that fall-through has any effect at all, then it means returning a user that is definitely not the same user object that the person who wrote the @-mention had selected. Very likely they have nothing at all to do with the person who was actually @-mentioned, and are just another person with the same name. That's no good -- much better to be honest about the fact that we simply didn't find the intended user.

In the context of the warning this code is part of providing, I guess the exact right thing to do is whatever the server is going to do to identify who's been @-mentioned. But I certainly hope that what the server is going to do when there's an explicit user ID is to stick to that, and not ignore it in favor of some potentially completely unrelated user who happens to have the same name. If not, then that's a bug that I expect would get fixed. So let's assume the server is doing the sensible thing, and do the same.

Reply:
Makes sense!